sweep/sweeper.go

```diff
@@ -1648,21 +1604,86 @@ func (s *UtxoSweeper) handleInputSpent(spend *chainntnfs.SpendDetail) {
 	}
 }
 
-// handleSweep is called when the ticker fires. It will create clusters and
-// attempt to create and publish the sweeping transactions.
-func (s *UtxoSweeper) handleSweep(bestHeight int32) {
+// markInputFailed marks the given input as failed and won't be retried. It
+// will also notify all the subscribers of this input.
+func (s *UtxoSweeper) markInputFailed(pi *pendingInput, err error) {
+	log.Errorf("Failed to sweep input: %v, error: %v", pi, err)
+
+	pi.state = StateFailed
+
+	// Remove all other inputs in this exclusive group.
+	if pi.params.ExclusiveGroup != nil {
+		s.removeExclusiveGroup(*pi.params.ExclusiveGroup)
+	}
+
+	s.signalResult(pi, Result{Err: err})
+}
+
+// updateSweeperInputs updates the sweeper's internal state and returns a map
+// of inputs to be swept. It will remove the inputs that are in final states,
+// and returns a map of inputs that have either StateInit or
+// StatePublishFailed.
+func (s *UtxoSweeper) updateSweeperInputs() pendingInputs {
+	// Create a map of inputs to be swept.
+	inputs := make(pendingInputs)
+
+	// Iterate the pending inputs and update the sweeper's state.
+	//
+	// TODO(yy): sweeper is made to communicate via go channels, so no
+	// locks are needed to access the map. However, it'd be safer if we
+	// turn this pendingInputs into a SyncMap in case we wanna add
+	// concurrent access to the map in the future.
+	for op, input := range s.pendingInputs {
+		// If the input has reached a final state, that it's either
+		// been swept, or failed, or excluded, we will remove it from
+		// our sweeper.
+		if input.terminated() {
+			log.Debugf("Removing input(State=%v) %v from sweeper",
+				input.state, op)
+
+			delete(s.pendingInputs, op)
+
+			continue
+		}
+
+		// If this input has been included in a sweep tx that's not
+		// published yet, we'd skip this input and wait for the sweep
+		// tx to be published.
+		if input.state == StatePendingPublish {
+			continue
+		}
```

Comment:
Nice catch - this will be clearer in PR 5 tho, where we peel off the publishing logic from the sweeper and create a tx publisher - it will make more sense to use this state there. But I do find a potential place we may end up stuck in this state in the final PR, need to think about it more..

Reply:
How can it get stuck in this state? IIUC, we have a single event loop in the sweeper as is. After something is marked as `StatePendingPublish`, then it can only go to published or failed as mentioned above. 

The only case I can see plainly is if the call to `StoreTx` fails (something something postgres txn issue), so then the transaction is published, but all inputs in that set are still marked as `StatePendingPublish`. In that case, we should mark the publish attempt as failed.